File: faucet/src/test/java/org/basinmc/faucet/util/VersionTest.java
/*
 * Copyright 2018 Johannes Donath  <johannesd@torchmind.com>
 * and other copyright owners as documented in the project's IP log.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"),
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.basinmc.faucet.util;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

import org.junit.jupiter.api.Test;

/**
 * @author <a href="mailto:johannesd@torchmind.com">Johannes Donath</a>
 * @since 1.0
 */
public class VersionTest {

  @Test
  public void testParse() {
    var version1 = new Version("0.0.0");
    var version2 = new Version("0.0.1");
    var version3 = new Version("0.1.0");
    var version4 = new Version("0.1.1");
    var version5 = new Version("1.0.0");
    var version6 = new Version("1.0.1");
    var version7 = new Version("1.1.0");
    var version8 = new Version("1.1.1");
    var version9 = new Version("1.1.1-alpha");
    var version10 = new Version("1.1.1-beta");
    var version11 = new Version("1.1.1-rc");
    var version12 = new Version("1.1.1-potato");
    var version13 = new Version("1.1.1+test");
    var version14 = new Version("1.1.1+abcdef");
    var version15 = new Version("1.1.1-alpha+abcdef");

    assertEquals(0, version1.major());
    assertEquals(0, version1.minor());
    assertEquals(0, version1.patch());
    assertFalse(version1.extra().isPresent());
    assertFalse(version1.buildMetadata().isPresent());

    assertEquals(0, version2.major());
    assertEquals(0, version2.minor());
    assertEquals(1, version2.patch());
    assertFalse(version2.extra().isPresent());
    assertFalse(version2.buildMetadata().isPresent());

    assertEquals(0, version3.major());
    assertEquals(1, version3.minor());
    assertEquals(0, version3.patch());
    assertFalse(version3.extra().isPresent());
    assertFalse(version3.buildMetadata().isPresent());

    assertEquals(0, version4.major());
    assertEquals(1, version4.minor());
    assertEquals(1, version4.patch());
    assertFalse(version4.extra().isPresent());
    assertFalse(version4.buildMetadata().isPresent());

    assertEquals(1, version5.major());
    assertEquals(0, version5.minor());
    assertEquals(0, version5.patch());
    assertFalse(version5.extra().isPresent());
    assertFalse(version5.buildMetadata().isPresent());

    assertEquals(1, version6.major());
    assertEquals(0, version6.minor());
    assertEquals(1, version6.patch());
    assertFalse(version6.extra().isPresent());
    assertFalse(version6.buildMetadata().isPresent());

    assertEquals(1, version7.major());
    assertEquals(1, version7.minor());
    assertEquals(0, version7.patch());
    assertFalse(version7.extra().isPresent());
    assertFalse(version7.buildMetadata().isPresent());

    assertEquals(1, version8.major());
    assertEquals(1, version8.minor());
    assertEquals(1, version8.patch());
    assertFalse(version8.extra().isPresent());
    assertFalse(version8.buildMetadata().isPresent());
  }

  @Test
  public void testCompare() {
    var versions = new Version[]{
        new Version("0.0.0"),
        new Version("0.0.1"),
        new Version("0.0.2"),
        new Version("0.1.0-alpha"),
        new Version("0.1.0-beta"),
        new Version("0.1.0"),
        new Version("0.1.1-alpha"),
        new Version("0.1.1-beta"),
        new Version("0.1.1"),
        new Version("0.2.0-alpha"),
        new Version("0.2.0-beta"),
        new Version("0.2.0"),
        new Version("1.0.0-alpha"),
        new Version("1.0.0-beta"),
        new Version("1.0.0"),
        new Version("1.0.1-alpha"),
        new Version("1.0.1-beta"),
        new Version("1.0.1"),
        new Version("1.1.0-alpha"),
        new Version("1.1.0-beta"),
        new Version("1.1.0"),
        new Version("2.0.0-alpha"),
        new Version("2.0.0-beta"),
        new Version("2.0.0"),
    };

    for (var i = 1; i < versions.length; ++i) {
      var version = versions[i];

      for (var j = 0; j < versions.length; ++j) {
        var other = versions[j];

        if (j < i) {
          assertTrue(version.isNewerThan(other), "version #" + i + " must be newer than #" + j);
          assertFalse(version.isOlderThan(other), "version #" + i + " cannot be older than #" + j);
          assertNotEquals(version, other, "version #" + i + " cannot be equal to #" + j);
          assertEquals(1, version.compareTo(other), "version #" + i + " must be larger than #" + j);
        } else if (j > i) {
          assertTrue(version.isOlderThan(other), "version #" + i + " must be older than #" + j);
          assertFalse(version.isNewerThan(other), "version #" + i + " cannot be newer than #" + j);
          assertNotEquals(version, other, "version #" + i + " cannot be equal to #" + j);
          assertEquals(-1, version.compareTo(other),
              "version #" + i + " must be larger than #" + j);
        } else {
          assertFalse(version.isNewerThan(other), "version #" + i + " cannot be newer than #" + j);
          assertFalse(version.isOlderThan(other), "version #" + i + " cannot be older than #" + j);
          assertEquals(version, other, "version #" + i + " must be equal to #" + j);
          assertEquals(0, version.compareTo(other), "version #" + i + " must be larger than #" + j);
        }
      }
    }
  }

  @Test
  public void testToString() {
    assertEquals("1.0.0", new Version("1.0.0").toString());
    assertEquals("1.0.1", new Version("1.0.1").toString());
    assertEquals("1.1.0", new Version("1.1.0").toString());
    assertEquals("2.0.0", new Version("2.0.0").toString());
  }
}

File: faucet/src/test/kotlin/org/basinmc/faucet/util/VersionTest.kt
/*
 * Copyright 2019 Johannes Donath <johannesd@torchmind.com>
 * and other copyright owners as documented in the project's IP log.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.basinmc.faucet.util

import org.junit.jupiter.api.Assertions.*

import org.junit.jupiter.api.Test

/**
 * @author [Johannes Donath](mailto:johannesd@torchmind.com)
 * @since 1.0
 */
class VersionTest {

  @Test
  fun testParse() {
    val (major, minor, patch, extra, buildMetadata) = Version("0.0.0")
    val (major1, minor1, patch1, extra1, buildMetadata1) = Version("0.0.1")
    val (major2, minor2, patch2, extra2, buildMetadata2) = Version("0.1.0")
    val (major3, minor3, patch3, extra3, buildMetadata3) = Version("0.1.1")
    val (major4, minor4, patch4, extra4, buildMetadata4) = Version("1.0.0")
    val (major5, minor5, patch5, extra5, buildMetadata5) = Version("1.0.1")
    val (major6, minor6, patch6, extra6, buildMetadata6) = Version("1.1.0")
    val (major7, minor7, patch7, extra7, buildMetadata7) = Version("1.1.1")
    val (major8, minor8, patch8, extra8, buildMetadata8, stability, stabilityIndex) = Version(
        "1.1.1-alpha")
    val (major9, minor9, patch9, extra9, buildMetadata9, stability1, stabilityIndex1) = Version(
        "1.1.1-beta")
    val (major10, minor10, patch10, extra10, buildMetadata10, stability2, stabilityIndex2) = Version(
        "1.1.1-rc")
    val (major11, minor11, patch11, extra11, buildMetadata11, stability3, stabilityIndex3) = Version(
        "1.1.1-potato")
    val (major12, minor12, patch12, extra12, buildMetadata12, stability4, stabilityIndex4) = Version(
        "1.1.1+test")
    val (major13, minor13, patch13, extra13, buildMetadata13, stability5, stabilityIndex5) = Version(
        "1.1.1+abcdef")
    val (major14, minor14, patch14, extra14, buildMetadata14, stability6, stabilityIndex6) = Version(
        "1.1.1-alpha+abcdef")

    assertEquals(0, major)
    assertEquals(0, minor)
    assertEquals(0, patch)
    assertNull(extra)
    assertNull(buildMetadata)

    assertEquals(0, major1)
    assertEquals(0, minor1)
    assertEquals(1, patch1)
    assertNull(extra1)
    assertNull(buildMetadata1)

    assertEquals(0, major2)
    assertEquals(1, minor2)
    assertEquals(0, patch2)
    assertNull(extra2)
    assertNull(buildMetadata2)

    assertEquals(0, major3)
    assertEquals(1, minor3)
    assertEquals(1, patch3)
    assertNull(extra3)
    assertNull(buildMetadata3)

    assertEquals(1, major4)
    assertEquals(0, minor4)
    assertEquals(0, patch4)
    assertNull(extra4)
    assertNull(buildMetadata4)

    assertEquals(1, major5)
    assertEquals(0, minor5)
    assertEquals(1, patch5)
    assertNull(extra5)
    assertNull(buildMetadata5)

    assertEquals(1, major6)
    assertEquals(1, minor6)
    assertEquals(0, patch6)
    assertNull(extra6)
    assertNull(buildMetadata6)

    assertEquals(1, major7)
    assertEquals(1, minor7)
    assertEquals(1, patch7)
    assertNull(extra7)
    assertNull(buildMetadata7)
  }

  @Test
  fun testCompare() {
    val versions = arrayOf(Version("0.0.0"), Version("0.0.1"), Version("0.0.2"),
        Version("0.1.0-alpha"), Version("0.1.0-beta"), Version("0.1.0"), Version("0.1.1-alpha"),
        Version("0.1.1-beta"), Version("0.1.1"), Version("0.2.0-alpha"), Version("0.2.0-beta"),
        Version("0.2.0"), Version("1.0.0-alpha"), Version("1.0.0-beta"), Version("1.0.0"),
        Version("1.0.1-alpha"), Version("1.0.1-beta"), Version("1.0.1"), Version("1.1.0-alpha"),
        Version("1.1.0-beta"), Version("1.1.0"), Version("2.0.0-alpha"), Version("2.0.0-beta"),
        Version("2.0.0"))

    for (i in 1 until versions.size) {
      val version = versions[i]

      for (j in versions.indices) {
        val other = versions[j]

        if (j < i) {
          assertTrue(version > other, "version #$i must be newer than #$j")
          assertFalse(version < other, "version #$i cannot be older than #$j")
          assertNotEquals(version, other, "version #$i cannot be equal to #$j")
          assertEquals(1, version.compareTo(other), "version #$i must be larger than #$j")
        } else if (j > i) {
          assertTrue(version < other, "version #$i must be older than #$j")
          assertFalse(version > other, "version #$i cannot be newer than #$j")
          assertNotEquals(version, other, "version #$i cannot be equal to #$j")
          assertEquals(-1, version.compareTo(other),
              "version #$i must be larger than #$j")
        } else {
          assertFalse(version > other, "version #$i cannot be newer than #$j")
          assertFalse(version < other, "version #$i cannot be older than #$j")
          assertEquals(version, other, "version #$i must be equal to #$j")
          assertEquals(0, version.compareTo(other), "version #$i must be larger than #$j")
        }
      }
    }
  }

  @Test
  fun testToString() {
    assertEquals("1.0.0", Version("1.0.0").toString())
    assertEquals("1.0.1", Version("1.0.1").toString())
    assertEquals("1.1.0", Version("1.1.0").toString())
    assertEquals("2.0.0", Version("2.0.0").toString())
  }
}
